### PR TITLE
fix: Bump hh deploy and solc versions

### PR DESCRIPTION
**Pull request by idea404**
# What :computer: 
* Bump hardhat zksync deploy and solc versions

# Why :hand:
* Fetches binaries from GH Releases CDN instead of repositories (related to `Headers Timeout Error`s found last week)

### CODE REVIEW

File: etc/system-contracts/package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
     "zksync-web3": "^0.13.0"
   },
   "devDependencies": {
-    "@matterlabs/hardhat-zksync-solc": "^0.3.15",
+    "@matterlabs/hardhat-zksync-solc": "^0.4.2",
```
Comment (MexicanAce) [inline etc/system-contracts/package.json:16]: I believe you'd also need to update `hardhat.config.ts` due to the following error: `Deprecated zksolc compiler version in hardhat config file (1.3.7). Please use versions 1.3.13 to 1.3.14.`

One thing to note, this directory is coming from https://github.com/matter-labs/era-system-contracts . So we can update for ourselves temporarily, but we should also consider fixing upstream at https://github.com/matter-labs/era-system-contracts/blob/main/package.json#L16

**Comment by idea404** [comment]
This PR reintroduces a function that was removed between the `hardhat-zksync-solc` versions `0.3.15` to `0.4.2`, namely this piece of code: 

```ts
import path from "path";
import { getCompilersDir } from "hardhat/internal/util/global-dir";
export async function getZksolcPath(version: string, salt: string = ""): Promise<string> {
  return path.join(await getCompilersDir(), "zksolc", `zksolc-v${version}${salt ? "-" : ""}${salt}`);
}
```

Since there is no formatting on `scripts` or this project level I skipped formatting the file.

File: e2e-tests/package.json [hunk 2]
```diff
@@ -5,8 +5,8 @@
   "author": "The Matter Labs Team <hello@matterlabs.dev>",
   "license": "MIT",
   "devDependencies": {
-    "@matterlabs/hardhat-zksync-deploy": "^0.6.3",
-    "@matterlabs/hardhat-zksync-solc": "^0.4.0",
+    "@matterlabs/hardhat-zksync-deploy": "^0.6.5",
+    "@matterlabs/hardhat-zksync-solc": "^0.4.2",
```
Comment (MexicanAce) [inline e2e-tests/package.json:9]: ah shoot, forgot to mention that we'll need to also update the `yarn.lock` file

Comment (idea404) [reply e2e-tests/package.json:9]: Doh should've known

**Review by MexicanAce — APPROVED** [review]
LGTM

**Pull request merged**